railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,383 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	"github.com/kong/kubernetes-ingress-controller/pkg/util"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(ctx context.Context, kongConfig sendconfig.Kong, log logr.Logger, kubeConfig *rest.Config, publishService string) {
+	ips, hostname, err := RunningAddresses(ctx, kubeConfig, publishService)
+	if err != nil {
+		log.Error(err, "failed to determine kong proxy external ips/hostnames.")
+		return
+	}
+
+	log.Info("Launching Ingress Status Update Thread.")
+
+	var wg sync.WaitGroup
+	for {
+		select {
+		case updateDone := <-kongConfig.ConfigDone:
+			log.V(4).Info("receive configuration information. Update ingress status %v \n", updateDone)
+			wg.Add(1)
+			go UpdateIngress(ctx, &updateDone, log, kubeConfig, &wg, ips, hostname)
```

Comment:
I think we already super clear about the data sync: function is synchronous, for sure. 
And we need go routine handle concurrency and avoid bottleneck coz a lot of ingresses comes in and need update. The  pointer parameter ensures that the exact ingress gets updated in that synchronous function.

Reply:
@rainest / @mflendrich could you please double check me on this? @ccfishk and I both talked about this at some length recently, but having another viewer double check it might save some time, appreciate it :bow:

Reply:
I think you're correct about the race condition in general, but not its effects: the contents of the channel events are config bundles only, and those don't include information about the proxy Service, only whether there's Kong config matching an Ingress in a config we successfully applied to the proxy.

IP information we retrieve with `RunningAddresses()` every time `PullConfigUpdate()` runs independent of the config bundle. Its result always reflects the most up-to-date IP information for the provided publish service.

The config may be out of date due to the race condition: you may, for example, emit a config event (configA), add an Ingress (Ingress1), and emit another config event (configB) that contains Ingress1's config. If your goroutine for configA runs after the goroutine for configB, the most recent set of status updates _will not_ include Ingress1.

There's an important caveat here though: as far as I know, we have no logic to _remove_ status information: we update all Ingresses we see in a given config, but if there's some other Ingress that we'd previously added status for and that is not in the last config we process, we simply do nothing with it. There is no list of Ingresses that we'd previously touched. This means that some Ingress may not receive a changed IP until another config event (rare, but possible) and that Ingresses we actually remove from config (e.g. by changing their class to something that no controller handles) will never lose their status info. The latter was true for 1.x also (it used a [list of current Ingresses in the store](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.2.0/internal/ingress/controller/controller.go#L168) and only cleared status on shutdown AFAIK).

IMO this is a relatively minor concern. It is a regression from 1.x (which instead ran status updates continuously independent of the sync loop, accessing the latest store contents each time) but overall nets us a benefit over 1.x (using the store meant that we could update status information on Ingresses that weren't actually available because of a config post failure).

The best option would probably to make this more like the 1.x loop, where status updates are continuous and pull the latest successful config independent of post events. This would require we add some way to retrieve config after we ship it, and IMO fits better into a broader refactor.

Reply:
Agree, the main purpose of the function is to ensure updating the exact status concurrently.

atm, we lack a response/confirm mechanism from Kong about what has been updated. We could cover that gap later on with more interface definition. would keep this implementation as it is.

Reply:
Thanks @rainest. I think given the conversation, I would be comfortable with us creating a follow up issue to refactor this later according to some of Travis' points, linking that issue here, and then considering this comment resolved. @ccfishk does that sound ok?

Reply:
Yeah, we can create follow up for optimization, which is good.

Reply:
I talked with @rainest in Slack about this and created the following followup for this to tighten up the loose ends:

https://github.com/Kong/kubernetes-ingress-controller/issues/1492

If you're satisfied with this followup @ccfishk, then please feel free to consider this comment _resolved_.

Reply:
close per suggestion.